### PR TITLE
multi-category support, rate limiting, and pagination

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This PR is a conglomerate of the following PRs.

https://github.com/opensearch-project/anomaly-detection/pull/60
https://github.com/opensearch-project/anomaly-detection/pull/64
https://github.com/opensearch-project/anomaly-detection/pull/65
https://github.com/opensearch-project/anomaly-detection/pull/67
https://github.com/opensearch-project/anomaly-detection/pull/68
https://github.com/opensearch-project/anomaly-detection/pull/69
https://github.com/opensearch-project/anomaly-detection/pull/70
https://github.com/opensearch-project/anomaly-detection/pull/71
https://github.com/opensearch-project/anomaly-detection/pull/74
https://github.com/opensearch-project/anomaly-detection/pull/75
https://github.com/opensearch-project/anomaly-detection/pull/76
https://github.com/opensearch-project/anomaly-detection/pull/77
https://github.com/opensearch-project/anomaly-detection/pull/78
https://github.com/opensearch-project/anomaly-detection/pull/79
https://github.com/opensearch-project/anomaly-detection/pull/82
https://github.com/opensearch-project/anomaly-detection/pull/83
https://github.com/opensearch-project/anomaly-detection/pull/84
https://github.com/opensearch-project/anomaly-detection/pull/92
https://github.com/opensearch-project/anomaly-detection/pull/94
https://github.com/opensearch-project/anomaly-detection/pull/93
https://github.com/opensearch-project/anomaly-detection/pull/95
https://github.com/kaituo/anomaly-detection-1/pull/1
https://github.com/kaituo/anomaly-detection-1/pull/2
https://github.com/kaituo/anomaly-detection-1/pull/3
https://github.com/kaituo/anomaly-detection-1/pull/4/
https://github.com/kaituo/anomaly-detection-1/pull/5
https://github.com/kaituo/anomaly-detection-1/pull/6
https://github.com/kaituo/anomaly-detection-1/pull/7
https://github.com/kaituo/anomaly-detection-1/pull/8
https://github.com/kaituo/anomaly-detection-1/pull/9
https://github.com/kaituo/anomaly-detection-1/pull/10

This spreadsheet contains the mappings from files to PR number (bug fix in my AD fork and tests are not included):
https://gist.github.com/kaituo/9e1592c4ac4f2f449356cb93d0591167

### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/85 (rate limiting)
https://github.com/opensearch-project/anomaly-detection/issues/87 (multiple category fields)
https://github.com/opensearch-project/anomaly-detection/issues/86 (pagination)
 
### Check List
- [ X ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java [hunk 1]
```diff
@@ -129,4 +143,35 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
...
+        if (Strings.isEmpty(detectorId)) {
+            throw new IllegalStateException(CommonErrorMessages.AD_ID_MISSING_MSG);
+        }
+
+        String entityName = request.param(CommonName.CATEGORICAL_FIELD);
+        String entityValue = request.param(CommonName.ENTITY_KEY);
+
+        if (entityName != null && entityValue != null) {
+            // single-stream profile request:
+            // GET _opendistro/_anomaly_detection/detectors/<detectorId>/_profile/init_progress?category_field=<field-name>&entity=<value>
```
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java:157]: nit: `_opendistro` -> `_plugins`

File: src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java [hunk 2]
```diff
@@ -129,4 +143,35 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
...
+            return Entity.createSingleAttributeEntity(detectorId, entityName, entityValue);
+        } else if (request.hasContent()) {
+            /* HCAD profile request:
+             * GET _opendistro/_anomaly_detection/detectors/<detectorId>/_profile/init_progress
```
Comment (ohltyler) [inline src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java:161]: nit: `_opendistro` -> `_plugins`

**Comment by ylwu-amzn** [comment]
Some high level question which not covered in previous PR review.

Current design will dispatch entities to different work nodes with consistent hashing. Each work node will cache entities in queue and run RCF model, then bulk write anomaly result to index. If current detection interval timeout, we just drop all remaining entities in queue to avoid carry over to next interval. It's possible that some worker node may receive more entities which needs to run longer like need to load checkpoints. So the unbalanced load may cause some worker node drops more entities. But seems not a big concern as we don't guarantee all entities will be detected, there are other reasons can cause dropping entities like worker node queue is full. 

Have some concern about the horizontal scalability. For example, if user scale cluster from 10 data nodes to 100 data nodes, then will have 100 worker nodes running entities, that means we may send out 100 bulk write AD result requests concurrently. But the AD result index shard setting is still the same, that may bring higher write load and user's general bulk write request may impacted. Confirm with PM/SDM if the concern of horizontal scalable issue is reasonable.

File: src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java:157]: fixed

File: src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/rest/RestGetAnomalyDetectorAction.java:161]: fixed

**Comment by kaituo** [comment]
> Some high level question which not covered in previous PR review.
> 
> Current design will dispatch entities to different work nodes with consistent hashing. Each work node will cache entities in queue and run RCF model, then bulk write anomaly result to index. If current detection interval timeout, we just drop all remaining entities in queue to avoid carry over to next interval. It's possible that some worker node may receive more entities which needs to run longer like need to load checkpoints. So the unbalanced load may cause some worker node drops more entities. But seems not a big concern as we don't guarantee all entities will be detected, there are other reasons can cause dropping entities like worker node queue is full.
> 
> Have some concern about the horizontal scalability. For example, if user scale cluster from 10 data nodes to 100 data nodes, then will have 100 worker nodes running entities, that means we may send out 100 bulk write AD result requests concurrently. But the AD result index shard setting is still the same, that may bring higher write load and user's general bulk write request may impacted. Confirm with PM/SDM if the concern of horizontal scalable issue is reasonable.

Will confirm with Sean about the horizontal scalability issue.  Result write is not problem, as we are gonna drop results when the index pressure is high.  But requests like reading checkpoints are impacted when
1)traffic is random (like ip address)
2)there are more entities than available memory can host

**Comment by kaituo** [comment]
> > Some high level question which not covered in previous PR review.
> > Current design will dispatch entities to different work nodes with consistent hashing. Each work node will cache entities in queue and run RCF model, then bulk write anomaly result to index. If current detection interval timeout, we just drop all remaining entities in queue to avoid carry over to next interval. It's possible that some worker node may receive more entities which needs to run longer like need to load checkpoints. So the unbalanced load may cause some worker node drops more entities. But seems not a big concern as we don't guarantee all entities will be detected, there are other reasons can cause dropping entities like worker node queue is full.
> > Have some concern about the horizontal scalability. For example, if user scale cluster from 10 data nodes to 100 data nodes, then will have 100 worker nodes running entities, that means we may send out 100 bulk write AD result requests concurrently. But the AD result index shard setting is still the same, that may bring higher write load and user's general bulk write request may impacted. Confirm with PM/SDM if the concern of horizontal scalable issue is reasonable.
> 
> Will confirm with Sean about the horizontal scalability issue. Result write is not problem, as we are gonna drop results when the index pressure is high. But requests like reading checkpoints are impacted when
> 1)traffic is random (like ip address)
> 2)there are more entities than available memory can host

Asked Sean. He said creating an issue to track https://github.com/opensearch-project/anomaly-detection/issues/125 and don't block code commit.

**Comment by kaituo** [comment]
@ylwu-amzn, please also comment on the issue I created with solutions.

**Review by ylwu-amzn — APPROVED** [review]

**Review by ohltyler — APPROVED** [review]

**Pull request merged**